internal/auth/token.go

```diff
@@ -0,0 +1,190 @@
+/*
+   Copyright 2024 Docker Terraform Provider authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package auth
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+)
+
+// TokenProvider provides a valid authentication token for API requests
+type TokenProvider interface {
+	// EnsureToken returns a valid token, refreshing if necessary
+	EnsureToken(ctx context.Context) (token string, err error)
+	// Username returns the username associated with this provider (for display purposes)
+	Username() string
+}
+
+// LoginTokenProvider uses static username/password to obtain tokens via login API
+// The name of this struct was specifically chosen
+// to troll iam-team :)
+type LoginTokenProvider struct {
+	username    string
+	password    string
+	baseURL     string
+	httpClient  *http.Client
+	token       string
+	tokenExpiry time.Time
+	mu          sync.Mutex
+}
+
+// NewLoginTokenProvider creates a token provider that uses username/password
+func NewLoginTokenProvider(username, password, baseURL string) *LoginTokenProvider {
+	return &LoginTokenProvider{
+		username:   username,
+		password:   password,
+		baseURL:    baseURL,
+		httpClient: http.DefaultClient,
+	}
+}
+
+func (p *LoginTokenProvider) EnsureToken(ctx context.Context) (string, error) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	// Return cached token if still valid
+	if p.token != "" && time.Now().Before(p.tokenExpiry) {
+		return p.token, nil
+	}
+
+	// Request new token
+	auth := struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+	}{
+		Username: p.username,
+		Password: p.password,
+	}
+
+	authJSON, err := json.Marshal(auth)
+	if err != nil {
+		return "", fmt.Errorf("marshal auth: %v", err)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", fmt.Sprintf("%s/users/login/", p.baseURL), bytes.NewBuffer(authJSON))
```

Comment:
issue: do not use `/users/login`. Instead use `/v2/auth/token`

**The deprecated login route will not accept org access tokens or JWTs**

https://docs.docker.com/reference/api/hub/latest/#tag/authentication-api/operation/AuthCreateAccessToken